### ISSUE TITLE
Week 2: Testing

### ISSUE THREAD

**Issue by des-des** [foundersandcoders/master-reference#16]
In the meeting on 9.12.16 (minutes in #10) we discussed tdd week. I do not think we managed to make a decision? Some points were:

 * Is qunit the right way to introduce fac to testing?
 * Do we need a tdd week?

The point that testing should be re-enforced throughout the course was made several times, I think this idea is already captured in the doc but we should be aware of this as we move forward.

**Comment by bradreeder** [comment]
I'm going to use this thread to update FAC9 feedback related to testing/tdd week if that's okay. Just to keep this in one discussion?

Few points from me:

* I don't think ES6 should be introduced until week 4+ alongside node to emphasise that they can use it in server-side code but be cautious on client-side.
* If it's possible to move linting to week 2 I wonder if it would be good as an introduction to good coding standards early on in the course?
* I agree with @iteles that we need both a week dedicated to testing and continual reinforcement throughout course.

**Comment by bradreeder** [comment]
We also need to come up with some kind of document that represents a clear community position on testing methodology (what to test, etc) for the purposes of the curriculum.

This is so that people don't contradict each other and confuse students throughout the course. This was a high priority issue with FAC9 #18 and something that came up continually as a source of confusion with FAC8.

**Comment by bradreeder** [comment]
FAC9 feedback #18:

* Introduce the concept of testing before tdd. Clear differentiation of testing as practice and tdd as methodology.
* If students are not expected to do everything tdd after the week, call it 'testing' week.
* Move committing against issues to week 2
* Decide the FAC line on testing and make sure that the mentor's intro on it is supported throughout the rest of the course (see my comment above.)
* Explain how to write testable code.
* Teach testing throughout course each time a new technology is introduced.
* Throughout course have projects that are tdd projects, and others that are not.

**Comment by jsms90** [comment]
@bradreeder Given the first bullet point, I suggest that we change this github issue to "Testing week" to set the tone as early as poss. Thoughts?

**Comment by bradreeder** [comment]
@jsms90 Introducing testing as practice followed by TDD as methodology makes sense. As long as the week introduces both and the project integrates both by encouraging a TDD approach I'm personally okay with the renaming.

**Comment by nelsonic** [comment]
@jsms90 _totally_ agree that the _emphasis_ should be on _testing_ in _general_ and the _label_ of `TDD` does not reflect the full dimension of what needs to be covered/understood.
The only thing I would _add_ is that TDD encourages writing tests _first_ which in turn means that _less_ code gets written. 👍

**Comment by jsms90** [comment]
I am all for TDD being the way that we encourage FAC students to write all of their code. Currently, that's not what we do. It may be unrealistic to expect this drastic a change to happen from one cohort to the next. Perhaps not.

Either way, I still think the terms are used interchangeably by my cohort, which in itself highlights the need for the name change.

**Comment by nelsonic** [comment]
@jsms90 F&C5 & 6 learned and practiced TDD throughout the course (_and beyond_)
ask @jrans, @simon, @des-des or @sohilpandya what effect that had on their later experience working on client work. Just as it's "_difficult_" to see the (_immediate_) benefit of digging a good foundation for a building, it pays off pretty quickly ... and _continues_ to pay through the life-span of the project. 
I don't _know_ why later F&C cohorts reverted back to "_bad habits_" ... hopefully we can address that in January by building something simple from "first principals" with tests. ❤️  ✅

**Comment by jsms90** [comment]
@nelsonic I agree that a good foundation pays off quickly. I feel very strongly about laying a solid foundation from the beginning for F&C10. All I meant to say is that it may be a bit much to expect a cohort who themselves have not really had experience of working in a TDD way, to then lead the next cohort in a purely TDD fashion. Unless this is something that can/will be changed in January.

I don't have a clue whether it is a true representation of the feeling amongst those in F&C, but it was very much my impression (at least from the first few weeks of the course), that very very few people in the building were pushing a TDD philosophy. Other than @sohilpandya ;) The belief that it is too difficult to employ whilst students are learning a new technology appears to be very common amongst mentors, as well as others from upstairs.

That's not to say that coding _should_ be done in a TDD way, just because I personally believe in the approach. Tbh I'm probably much too inexperienced to make that kind of overarching judgement. But it doesn't surprise me at all that F&C9 haven't adopted it. If it's something that F&C believes in as strongly as pair programming, for example, that didn't come across at all. It came across much more like a recommendation, at best.

**Comment by nelsonic** [comment]
@jsms90 yeah, I _get_ it. 👍 
I don't _know_ if TDD is something that F&C believes in ... because I don't know what F&C _believes_ in. Writing _reliable_ (_aka testable_) code is not a "_nice to have_" _suggestion_ ... thought _most_ agencies and businesses people are likely to encounter consider testing to be an _after thought_ because of an _out-dated_ "[_Waterfall_](https://en.wikipedia.org/wiki/Waterfall_model)" (_all requirements up front. build. then test, maybe._) mindset. 

# tl;dr

Perhaps the most important _concept_ people are not being introduced to is the "spike".
https://en.wikipedia.org/wiki/Spike_(software_development)

When you are learning something for the _first_ time it can be _unrealistic_ to write a _test_ `before` you know what you are _trying_ to build ... so, you _first_ write some "trying a few things to see what works" code to learn what an API is or how to us a particular tool and once you have a good grasp of what's _possible_, you _scrap_ the "working but untested" code and write your first test. At this point you have a fairly _decent_ idea of what _challenge_ you are attempting to solve and have an insight into _how_ you _can_ go about testing for it.

> Maybe the word "_testing_" is _inappropriate_/_misleading_/_confusing_ and instead we should refer to the practice as: "_truth-finding_" i.e. what is _true_ in our application. 
This would _explain_ why our _assertions_ are usually testing for truth (_or equality_).
At the _core_ level the `assert` module does _exactly_ this. https://nodejs.org/api/assert.html#assert_assert_value_message
the `assert` _methods_ are just "_syntactic sugar_" on top of this "truth" check.
`then` the act of "_testing_" the application is simply to discover what the "truths" are. 
e.g: `assert(add(1, 1) === 2, 'add two numbers returns the sum of the two arguments');` ... 
we can _deduce_ what the `add` method needs to do here. _right_?
Then it's the case of building up people's _confidence_ in the "_truths_" they are seeking from their methods `while` ensuring that the _previously_ established "_truths_" are not "_broken_" in the process of creating new "_truths_"... 🤔 

For the _curious_/_conscientious_ I'd _highly_ recommend reading: https://www.amazon.co.uk/Growing-Object-Oriented-Software-Guided-Signature/dp/0321503627
While I would not _advocate_ the "OO" design pattern necessarily and the book is written in Java, it's the _best_ book for explaining why organically "_growing_" your codebase through test-driven iteration is the _only_ way to produce maintainable/reliable code.

My _plan_ has _always_ been to "_translate_" the Java-world books into JS ... so others don't have to go through the _pain_ & _complexity_ of the OO pattern only to discover _years_ into their "_career_" that "FP" is _much_ simpler (_to "reason about" than OO_) once you give it a chance ... 

I'm _truly sorry_ that the book hasn't been written yet. 😢 
We got _side-tracked_ writing code for clients in 2016 which produced a number of popular tutorials (_with tests_):
+ https://github.com/dwyl/learn-aws-lambda
+ https://github.com/dwyl/learn-nightwatch
+ https://github.com/dwyl/learn-redux
+ https://github.com/dwyl/learn-graphQL ... etc.
The _plan_ is to re-start it in January 2017 and "dog-food" it on an [_ambitious_](https://github.com/dwyl/hq/issues/128) project that _solves a real human challenge_.
So there will be _end-to-end_ examples which people can learn with and contribute components to a _real_ application used by (_potentially hundreds of thousands of_) people ...

**Comment by sohilpandya** [comment]
I'm sorry to be a bit off topic but I'm well impressed that both of you referred to founders and Coders as F&C and not FAC. 🎉

On a slightly serious note, I've put myself forward for helping out with planning the Testing and Postgres week. So I'll definitely be giving my two cents worth to those who pick that week up! 

☺️

**Comment by jrans** [comment]
Hey! My few thoughts on testing. I'm not a die hard fan of any philosophy but just wanted to highlight some of what I think are the most important aspects in my coding journey. Note I have worked as both a junior member, lead developer and code reviewer or a project with increasing experience and so have different perspectives. I have missed things out/waffled on about the obvious but I hope at least one person finds some of this useful!

+ TDD is a great way to learn testing. Not only do you start to structure your code differently but it completely breaks down the problem into small manageable pieces that become very simple to solve and can really steer development. Your tests clearly represent what your application/function does and you should expect no more/less from it. Not only do you think about all the decisions towards creating a solution but you are left with a document of that that can forever keep the integrity of your application.
+ I love unit testing. Simply because it breaks down a problem and code code into small manageable pieces that can be reusable in other parts of the application or other applications all together. *"Take care of your pennies and your pounds will take care of themselves"* and all that jazz. Obviously you still need to have a good overview of the problem you're trying to solve and be able to see clearly how thing will fit together but once you have an idea of a path take things one step at a time. Unit testing encourages smaller functions that focus on one thing and are therefore very easy to test. Say you have 5 steps to your process, each of them having 2 branches. That gives potentially 2*5 = 32 possible outcomes that you would have to consider if you were testing the whole process. Alternatively if you are confident each step is good and you've glued them together logically then you could instead have 5 * 2 + 1 integration test  = 11 and you'd be far more certain you;d have covered every case. That ties in with my next point:
+ Unit testing pin points bugs. I instantly no what part of the code to go to make the fix. Simple. Tests of increasing granularity will always help with this.
+ No one likes writing tests and wants to keep it to a minimum hence write more elegant simple code! If something is hard to test then it could probably be done more simply. If you stick with the strict rule that you want all code completely tested and covered but you want to reduce this task it will only end up forcing you to write less, more efficient, mindful and reusable code which is winning all around. If you find yourself writing the same test again and again you're probably writing the same code again and again.
+ Testing is very important if you value your work and you are contributing in a team. If you include your tests into the project and people want to hack away at your code the one way to keep posterity is to make them consciously change your tests too. If your function stops working randomly, yet your tests are passing then you haven't properly tested but if you have then you'll avoid the situation altogether. Tests are your insurance policy, cover yourself!
+ Related to previous points but I'm still amazed at how having to test something has completely shaped by coding style and avoided writing bad code. I can instantly smell a stink when I know something will be hard to test, i.e.  having a function with 5 nested callbacks. Be grateful to the process!
+ I rarely TDD in the strict sense. However I achieve the same results by following a very rigid process that leaves little room for error. Committing little and often on the condition of 100% coverage means that I never go too long coding without tests. I like the freedom of exploring a problem and iterating on my solution quickly but I remain mindful of how I'm going to test. I then use the testing as a way to proof read my code and refactor! It really works for me. I'm amazed at how many mistakes/typos I make and I instantly uncover them without having to run my whole application again and again. Saved my skin a lot of times and again left with tests to save it again and again in the future when I make modifications. That is one thing that really annoys me when reviewing code is people so often just press submit PR and haven't even proof read there diffs! Wastes everyones time but most importantly mine ;).
+ So TDD isn't adopted on my day to day basis when I'm fairly comfortable with what I'm doing however it finds its element again as soon as I'm lost! When I have no clue how to solve something I instantly adopt the practise for all the reasons stated above! So there you go, great for learning how to test but also great to come back to in times of need.
+ Unit testing is great for development of a project and fuelling the whole process, however integration testing is there for the end and helps maintain a large code base over time. Sadly all my projects have been relatively young and so I haven't explored integration testing as much as I'd have liked to. I do completely value it but I feel its most necessary when you have finished rapidly iterating and want to keep hold of a functioning app. Again it pains me to ever have to click through screens before a client demo again and again and I feel there should be a focus to make an easy solution to this in the space!

**Comment by bradreeder** [comment]
Week 2 S-G-C notes:

##### Stop

* 'Doing TDD wrong'
* Moving goalposts - i.e. pick one framework for the week and stick to it. 'here's jasmine -- you can't use it.'

##### Go

* Resources tab on gitbook to include (article?) on differences between frameworks e.g. Jasmine & QUnit (if you use two frameworks this week).
* More on async functions precourse.
* 5 minute breaks
* Introduce modularisation e.g. through workshop
* Reverse order of FizzBuzz & dwyl workshop

##### Continue

* Emphasise starting projects with design phase.
* Research day every week.
* TDD on all projects (interesting :P)
* Heads up on next week's topics - not compulsory homework.
* Swapping pairs
* Learning git and TDD through own projects
* Pushing work up to github regularly.

**Comment by des-des** [comment]
@nelsonic, I think your point about doing a spike _before_ writing production code is related to the discussion around readme day #17. In my mind, on tuesday students should write some code, understand how to test it, document their findings, then throw it away and get ready to start the project.

This point is more generally about testing through out the course, a _documented_ spike (then some tests) on Tuesdays may both be a solution to broken readme day and start to fill the gap in testing knowledge in F&C.

**Comment by RhodesPeter** [comment]
# Planning

## Learning Objectives

- What is testing? Why do we do it?
  - Unit & integration tests
- What is TDD?
  - Red Green Refactor cycle
- How to use Qunit.
  - How to test asynchronously

## Project
 - Stopwatches - in pairs or groups of 4?

## Resources
 - https://github.com/dwyl/learn-tdd
 - https://github.com/indykid/romanizer_js
 - https://github.com/JMurphyWeb/testing-morning-challenge
 - http://codingdojo.org/cgi-bin/index.pl?KataCatalogue

## README topics
 - Unit vs Integration testing?
 - TDD lifecycle
 - Continuous integration
 - QUnit
 - Technical Spikes

## Schedule

### Monday
  - 10:00 - 11:00
      - Intro to testing
      - Fizz Buzz
  - 11:00 - 13:00
      - [DWYL TDD workshop](https://github.com/dwyl/learn-tdd) - This may pass over into the afternoon
  - 14:00 - 16:00
      - Ping pong roman numeral challenge pair programming
  - 16:00 - 18:00
      - Business development / community engagement 

### Tuesday
  - 10:00 - 11:00
    - Morning Challenge - Callbacks (we will create/borrow/steal one)
  - 11:00 - 13:00
    - Intro to async, testing async and testing things like stopwatches 😉
  - 14:00 - 18:00
    - READMEs 

### Wednesday
   - 10:00 - 11:00
      - morning challenge - waterfall functions
   - 11:00 - 18:00
      - Start projects 

### Thursday
- 10:00 - 18:00
   - Group projects  

### Friday
  - 10:00 -11:00
    - Code review
  - 11:00 - 13:00
     - Respond to issues
  - 14:00 - 14:45
     - Group stop-go-continue
   - 14:45 - 17:00
     - Present projects
   - 17:00 - 18:00
     - Team stop-go-continue

**Comment by Jwhiles** [comment]
I've been working on a little workshop/tutorial to help the (young FACs write their programs in a functional manner.)[https://github.com/Jwhiles/pure-functions-easy-testing]

I'm going to add some exercises, more examples and, and rewrite it for clarity. I think this is part of writing testable code that is important to learn, and that we didn't really get during testing week.

Thoughts?  :)

**Comment by lucymonie** [comment]
I have a potential speaker for you guys, which I discussed briefly with @RhodesPeter. It's @hdrdavies (alumni). He could do a talk on testing 5-6pm on Tuesday or Thursday of testing week. If this seems good, let me know and we can confirm a date. It's TBC in the FAC calendar.

**Comment by skibinska** [comment]
My suggestions:

**Weekend resources**
* [The Cycles of TDD](http://blog.cleancoder.com/uncle-bob/2014/12/17/TheCyclesOfTDD.html)
* [Test-driven development](https://en.wikipedia.org/wiki/Test-driven_development)


**Learning outcomes**
1. What is testing? Why do we do it?
    * Unit & integration tests
2. What is TDD?
   * Red Green Refactor cycle
3. How to use QUnit?
   * How to set up QUnit
   * How to test synchronously
   * How to test asynchronously


**README topics**

* **Unit vs Integration testing**
    * What are the differences
    * What are the pros and cons
    * Examples of each
* **Technical Spikes**
  * What is it ?
  * Why is it helpful ?

I can't think of any other README topics. 
I did.t include the following because:
* Continuous integration - this topic is covered in [node week 5](https://github.com/foundersandcoders/master-reference/issues/31)
* QUnit - I don't really see the point to make it a README topic as we will cover it in our workshops
* TDD lifecycle - as above - it is already in [dwyl tutorial](https://github.com/dwyl/learn-tdd)

@bradreeder , @Jwhiles, @RhodesPeter any suggestions?

**Comment by RhodesPeter** [comment]
@lucymonie yes please! Could you please ask Huw Davies if he could come in at 4pm on Thursday? We have penciled him in :)

**Comment by skibinska** [comment]
I would like to add [this](https://medium.com/javascript-scene/what-every-unit-test-needs-f6cd34d9836d#.wcbhgv3ue) article on 5 Questions Every Unit Test Must Answer to Weekend resources

**Comment by skibinska** [comment]
I still don't know what Readme topics should be. Here is a list of my suggestions:

* BlanketJS
* PhantomJS
* manual testing vs automation testing

Any thoughts?

**Comment by jsms90** [comment]
Any reason why Huw Davies is on Thursday rather than Wednesday?

**Comment by RhodesPeter** [comment]
Hew can only do Tuesday or Thursday.

**Comment by jsms90** [comment]
Gotcha!

**Comment by hdrdavies** [comment]
Hi! (it's Huw)

**Comment by RhodesPeter** [comment]
Hi Huw. Thank you for agreeing to give us a talk during testing week. Lucy passed on your email so I will touch base with you there. If you have any questions feel free to ask them here or via email.

**Comment by sohilpandya** [comment]
@/all Thanks again for all the great work! it's coming up to the time where you'll all be busy working on projects. What's the current status of this week? Do we have all the content + resources in place ready to go? We don't want to be in a position where we are planning it the week before 😟 

Might be worth everyone getting together, finalising the details soon so that are ready! 😄 Happy to be present for it.

**Comment by sohilpandya** [comment]
@/all did you guys get a chance to meet up and finalize the details/resources? I know you've had a super busy week and it's probably going to be the same this week, but putting some time aside for sorting this out at this stage would also be very helpful! 😄

**Comment by sohilpandya** [comment]
@/all Hope you are all doing well, last week prior to FAC10, time for you guys to reconvene and finalize the details?

I really don't want this to become something that is done at the last minute in week 1.

**Comment by skibinska** [comment]
hi @sohilpandya, I would appreciate if you could help me with choosing README topics.  
Here are my suggestions:

* **Unit vs Integration testing**
    * What are the differences
    * What are the pros and cons
    * Examples of each
* **Technical Spikes**
  * What is it?
  * Why is it helpful?
* **BlanketJS**
   * What is it?
  * Why is it helpful?
* **PhantomJS**
    * What is it?
    * Why is it helpful?
* **manual testing vs automation testing**
    * What are the differences
    * What are the pros and cons
    * Examples of each
* **QUnit Assertions (other than equal())** - in more details, examples
* **SauceLabs Cross-Browser Testing**
    * What is it?
    * Why is it helpful?

**Comment by sohilpandya** [comment]
@skibinska looks great! Thanks for the hard work! I just have a couple of questions:
- Why are we going with BlanketJS and PhamtomJS? Is it worth staying with Travis, codeClimate, circleCI?

We could add another topic:
- What is TDD
  - TDD Lifecycle
  - Benefits of TDD 
  - Red Green Refactor

It might be worth asking them for a real-life example for all of the topics that we list so that they have an end goal for the day, they can then ask the other members to walk through the tutorial alongside reading the readme. 

Thanks again! 😄

**Comment by hdrdavies** [comment]
Could do with a chat with someone to figure out what sort of stuff you guys would like me to cover too. Next week is good. 👍

**Comment by skibinska** [comment]
@sohilpandya I didn't include the following because:
* Continuous Integration (TravisCI, Codecov, Code Climate) - this is a README topic in [node week 5](https://github.com/foundersandcoders/master-reference/issues/31)
* TDD - [dwyl workshop](https://github.com/dwyl/learn-tdd) covers it (we are planning to do it on Monday).

I think during our cohort someone mentioned BlanketJS, that's way I added it.

>  It might be worth asking them for a real-life example for all of the topics that we list

:+1:  really like this idea!

**Comment by sohilpandya** [comment]
Great! how about the rest of the schedule for the week? is it ready to go?

**Comment by skibinska** [comment]
hi /all, are we agree to keep these Readme topics : 

* **Unit vs Integration testing** (4 people)
    * What are the differences
    * What are the pros and cons
    * Examples of each
* **Manual testing vs Automation testing** (4 people)
    * What are the differences
    * What are the pros and cons
    * Examples of each
* **Technical Spikes** (2 people)
  * What is it?
  * Why is it helpful?
  * Example
* **QUnit Assertions (other than equal())** - in more details, examples (2 people)
* **BlanketJS** (4 people)
   * What is it?
   * Why is it helpful?
   * Example of it

**Comment by sohilpandya** [comment]
@skibinska I'm good with that, anyone else have any thoughts?

**Comment by RhodesPeter** [comment]
Fine by me @skibinska :)

**Comment by skibinska** [comment]
hi @all, could give me some feedback on [qunit intro](https://github.com/skibinska/learn-qunit/blob/master/README.md) ?

**Comment by skibinska** [comment]
hello again, here is my first attempt on [fizzbuzz workshop](https://github.com/skibinska/fizzbuzz).
@all, @sohilpandya  please let me know what you think!

**Comment by skibinska** [comment]
after speaking to @sohilpandya, I did a small change in readme topics - removed manual testing vs automation testing.

**Comment by bradreeder** [comment]
Week delivered. Thanks everyone!
